src/rest_vol_dataset.c

```diff
@@ -4526,3 +4606,90 @@ RV_json_values_to_binary_callback(char *HTTP_response, void *callback_data_in, v
+
+    if ((dtype_class = H5Tget_class(dtype_id)) == H5T_NO_CLASS)
+        FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL, "can't get datatype class");
+
+    if (dtype_class == H5T_INTEGER) {
+        if (H5Tequal(dtype_id, H5T_NATIVE_INT) != TRUE)
+            FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
+                            "parsing non-native integer types is unsupported");
+
+        if (!YAJL_IS_INTEGER(value_entry))
+            FUNC_GOTO_ERROR(H5E_OBJECT, H5E_PARSEERROR, FAIL,
+                            "parsed yajl val has incorrect type; expected integer");
+
+        *((int *)value_buffer) = (int)YAJL_GET_INTEGER(value_entry);
+    }
+    else if (dtype_class == H5T_FLOAT) {
+        if (H5Tequal(dtype_id, H5T_NATIVE_FLOAT) == TRUE) {
+            if (!YAJL_IS_DOUBLE(value_entry))
+                FUNC_GOTO_ERROR(H5E_OBJECT, H5E_PARSEERROR, FAIL,
+                                "parsed yajl val has incorrect type; expected float-like");
+
+            *((float *)value_buffer) = (float)YAJL_GET_DOUBLE(value_entry);
+        }
+        else if (H5Tequal(dtype_id, H5T_NATIVE_DOUBLE) == TRUE) {
+            if (!YAJL_IS_DOUBLE(value_entry))
+                FUNC_GOTO_ERROR(H5E_OBJECT, H5E_PARSEERROR, FAIL,
+                                "parsed yajl val has incorrect type; expected double");
+
+            *((double *)value_buffer) = YAJL_GET_DOUBLE(value_entry);
+        }
+        else {
+            FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
+                            "parsing non-native float types is unsupported");
+        }
+    }
+    else if (dtype_class == H5T_COMPOUND) {
+        /* Recursively parse each member of the compound type */
+        int      nmembers        = 0;
+        size_t   offset          = 0;
+        size_t   member_size     = 0;
+        hid_t    member_dtype_id = H5I_INVALID_HID;
+        yajl_val member_val;
+
+        if ((nmembers = H5Tget_nmembers(dtype_id)) < 0)
+            FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL,
+                            "can't get number of members in compound datatype");
+
+        for (int i = 0; i < nmembers; i++) {
+            if ((member_dtype_id = H5Tget_member_type(dtype_id, (unsigned int)i)) < 0)
+                FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL,
+                                "can't get datatype of member in compound datatype");
+
+            if ((member_size = H5Tget_size(member_dtype_id)) == 0)
+                FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL, "can't get size of member datatype");
+
+            if ((member_val = YAJL_GET_OBJECT(value_entry)->values[i]) == NULL)
+                FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_PARSEERROR, FAIL,
+                                "failed to parse member of compound type");
+
+            if (RV_json_values_to_binary_recursive(member_val, member_dtype_id,
+                                                   (void *)((char *)value_buffer + offset)) < 0)
```

Comment:
This will probably work for packed compound types, but if the compound datatype has padding in it in the same way that the natural C struct likely does, this will be problematic. Best to use H5Tget_member_offset to get the member offsets instead of assuming they're linearly next to each other